File: strings/generateDocument.js
// my solution 9/20/23
function generateDocument(characters, document) {
  // create charObj
  const charObj = {};
  // create docObj
  const docObj = {};
  // for loop to add characters and their amount to charObj
  for (let char of characters) {
    if (charObj[char] === undefined) {
      charObj[char] = 1;
    } else {
      charObj[char]++;
    }
  }
  // same for docObj
  for (let char of document) {
    if (docObj[char] === undefined) {
      docObj[char] = 1;
    } else {
      docObj[char]++;
    }
  }
  // for in loop through docObj
  for (let key in docObj) {
    // if docObj[key] is not greater than or equal to charObj[key]
    if (docObj[key] > charObj[key] || charObj[key] === undefined) {
      // return false
      return false;
    }
  }
  return true;
}

File: strings/generateDocument.py
def generateDocument(characters, document):
    # create charDict
    charDict = {}
    # for loop over char to determine if in charDict, if not add it, otherwise, increment
    for i in range(len(characters)):
        currChar = characters[i]
        if currChar not in charDict:
            charDict[currChar] = 1
        else:
            charDict[currChar] += 1

    # for loop over document
    for j in range(len(document)):
        currDocChar = document[j]
        # if document[char] is not there, or equal to zero, then return False
        if currDocChar not in charDict or charDict[currDocChar] == 0:
            return False
        charDict[currDocChar] -= 1

    return True
